ConnectorRevit/ConnectorRevit/UI/ConnectorBindingsRevit2/ConnectorBindingsRevit2.Mappings.cs

```diff
@@ -0,0 +1,515 @@
+    /// <param name="sourceApp"></param>
+    /// <returns></returns>
+    public async Task UpdateForCustomMapping(StreamState state, ProgressViewModel progress, string sourceApp)
+    {
+      // Get Settings for recieve on mapping 
+      var receiveMappingsModelsSetting = (CurrentSettings.FirstOrDefault(x => x.Slug == "receive-mappings") as MappingSeting);
+      var receiveMappings = receiveMappingsModelsSetting != null ? receiveMappingsModelsSetting.Selection : "";
+
+      Dictionary<string, List<MappingValue>> settingsMapping = null;
+      bool isFirstTimeReceiving = false;
+
+      if (receiveMappingsModelsSetting.MappingJson != null)
+        settingsMapping = JsonConvert.DeserializeObject<Dictionary<string, List<MappingValue>>>(receiveMappingsModelsSetting.MappingJson);
+      else
+        isFirstTimeReceiving = true;
+
+      if (receiveMappings == noMapping || receiveMappings == null)
+        return;
+      else
+      {
+        Dictionary<string, List<string>> hostTypesDict = GetHostTypes();
+        Dictionary<string, List<string>> incomingTypesDict = GetIncomingTypes(progress, sourceApp);
+
+        // if mappings already exist, update them and return true
+        bool newTypesExist = UpdateExistingMapping(settingsMapping, hostTypesDict, incomingTypesDict, progress);
+
+        Dictionary<string, List<MappingValue>> Mapping = settingsMapping;
+        if (Mapping == null)
+          Mapping = returnFirstPassMap(incomingTypesDict, hostTypesDict, progress);
+
+        try
+        {
+          // show custom mapping dialog if the settings corrospond to what is being received
+          if (newTypesExist || receiveMappings == everyReceive)
+          {
+            var vm = new MappingViewModel(Mapping, hostTypesDict, progress, newTypesExist && !isFirstTimeReceiving);
+            MappingViewDialog mappingView = new MappingViewDialog
+            {
+              DataContext = vm
+            };
+
+            Mapping = await mappingView.ShowDialog<Dictionary<string, List<MappingValue>>>();
+
+            receiveMappingsModelsSetting.MappingJson = JsonConvert.SerializeObject(Mapping); ;
+
+          }
+          // update the mapping object for the user mapped types
+          SetMappedValues(Mapping, progress, sourceApp);
+        }
+        catch (Exception ex)
+        {
+          progress.Report.Log($"Could not make new mapping {ex}");
+        }
+      }
+    }
+
+    /// <summary>
+    /// This method creates a mapping dictionary by looping through all of the incoming types and mapping them to the most similar host type of the same category
+    /// </summary>
+    /// <param name="incomingTypesDict"></param>
+    /// <param name="hostTypes"></param>
+    /// <param name="progress"></param>
+    /// <returns></returns>
+    public Dictionary<string, List<MappingValue>> returnFirstPassMap(Dictionary<string, List<string>> incomingTypesDict, Dictionary<string, List<string>> hostTypes, ProgressViewModel progress)
```

Comment:
Very minor remark #2: function names should be camel case, and they should be private or internal instead of public, unless necessary ofc